network/src/connectivity_manager/mod.rs

```diff
@@ -130,27 +130,57 @@ struct DialState<TBackoff> {
     addr_idx: usize,
 }
 
-impl<TTicker, TBackoff> ConnectivityManager<TTicker, TBackoff>
+/// The configuration fields for ConnectivityManager
```

Comment:
what's the motivation for splitting this? I actually removed a lot of these "config" structures from consensus because it felt like I was jumping through hoops to instantiate the service(s).

Reply:
1.  Hierarchically apply the Builder pattern at every level.  main_node can be viewed as node_builder and needs to glue network/consensus/statesync/mempool together to form the node.  network_builder needs to glue discovery, peermanager, connectivitymanager, ... together to form the network.
1.  Each builder should have a complete configuration which is sufficient to build the underlying component.  This may require some amount of logic (i.e., setup_network should actually be a part of NetworkBuilder, and is logic applied to configuring internal components and wiring them together.
1.  Overwhelming majority of the 'shared' fields in network_builder can be removed/moved into the component specific configs so network_builder is a high level view of the system (which components exist) and is not tied to internal details of those components
1.  A fair number of the hoops and circular dependencies are tied to configuring/building/starting the components all at once.   Having the configs does two things.  First, it allows for isolating configure/build/start into separate passes (helpful for dealing with dependencies).  Second,  it makes components look the same regardless of where they sit in the system hierarchy and helps to enforce abstraction isolation.

Caveat:  explicit config splitting is most useful if the configure/build/start phases are explicitly split from each other.  if they bleed together then the configs simply exacerbate the already existing problem.